More benchmarker features

rctcwyvrn opened the pull request:
Adds 
- Loading results from a file
- Measuring compile times (via a spi function in `Regex`)
- Comparing compile times and against NSRegularExpression
- Saving comparison results as csv
- Charts (based on #585)

rxwei reviewed (approved):
📊

rxwei commented on Sources/RegexBenchmark/BenchmarkChart.swift:
```diff
@@ -0,0 +1,81 @@
+    sampleCount: Int
+  ) -> some ChartContent {
+    // Normalized runtime
+    BarMark(
+      x: .value("Name", name),
+      y: .value("Normalized runtime", new / old))
+    .foregroundStyle(LinearGradient(
+      colors: [.accentColor, new - old <= 0 ? .green : .yellow],
+      startPoint: .bottom,
+      endPoint: .top))
```
You could inline this into the `ForEach { ... }`. I pulled this out to `chartBody` because of "expression too complex". That doesn't happen with just one bar mark.

rctcwyvrn commented:
@swift-ci test

rctcwyvrn commented:
Oh huh what version is the CI on?

```
/Users/ec2-user/jenkins/workspace/pr-swift-experimental-string-processing-macos/branch-main/swift-experimental-string-processing/Sources/RegexBenchmark/BenchmarkChart.swift:14:8: error: no such module 'Charts'
import Charts
       ^
```

rxwei commented:
I don't think the CI runs on macOS 13. Maybe `#if canImport(Charts)` around the import and around `Chart` usage.

milseman commented on Sources/_StringProcessing/Regex/Core.swift:
```diff
@@ -135,6 +135,25 @@ extension Regex {
+      return false
+    }
+  }
+}
```
This kind of approach looks ok to me. I wonder if we can generalize the interface a little bit to give us some future flexibility without needing to add more SPI over time. For example, we could return a (resilient) result type which we can add metrics to in the future (number of instructions, complexity or star-height, etc). We could also have a (resilient) compilation options parameter, such as the proposed force-recompilation behavior which we could extend with more options like control over optimizations or even targeted engine version (e.g. back deployment)

rctcwyvrn commented:
- Added setting compiler options to the SPI interface
- Added metrics and tracing enabled as options for the compiler 
- Using the new additions, added CLI options for turning on metrics/tracing for benchmarker regexes
- Added parse time measurements to the benchmarker

rctcwyvrn commented:
@swift-ci test

rctcwyvrn commented:
@milseman Could you check that `Regex._forceAction` in `Core.swift` matches what you had in mind? Also that it's ABI stable the way it's written right now?

milseman commented on Sources/_StringProcessing/Engine/Processor.swift:
```diff
@@ -89,6 +89,8 @@ struct Processor {
   // MARK: Metrics, debugging, etc.
   var cycleCount = 0
   var isTracingEnabled: Bool
+  let shouldMeasureMetrics: Bool
```
While well-predicted, do we see any impact from checking this bool inside the processor? Worst case we could make it conditionally compile and the benchmark can enable it.

rctcwyvrn replied:
<img width="907" alt="image" src="https://user-images.githubusercontent.com/49693777/184702070-6cbd65e5-c4fa-4f4a-9d74-062834ef2e41.png">

Ah... the answer is quite a lot. Removing both the tracing and metric conditions speeds things up by 10-20% across the board. I'll see if I can figure out the conditional compilation

rctcwyvrn replied:
After looking around it doesn't seem like swiftPM supports the feature of "define this compilation variable only when being built as a dependency for target X". The only conditions it can enable/disable compilation variables on are platforms and debug/release. Trying to add multiple targets with flags enabled/disabled for `_StringProcessing` ends up with "overlapping source code" 

https://forums.swift.org/t/swiftpm-and-conditional-compilation/36874
https://forums.swift.org/t/compilation-conditions-and-swift-packages/34627/4

milseman reviewed (approved):
LGTM after making sure to underscore the SPI

rctcwyvrn commented:
<img width="1424" alt="image" src="https://user-images.githubusercontent.com/49693777/184707140-e375e27e-97fc-4584-861e-18c5e33eb22e.png">

Made the charts nice and pretty <:

rctcwyvrn commented:
@swift-ci test

rctcwyvrn commented:
@swift-ci test

rctcwyvrn commented:
@swift-ci test

rctcwyvrn commented:
augh I forgot to switch it back to squash... oops
